core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala

```diff
@@ -0,0 +1,191 @@
+import kafka.server.LogOffsetMetadata
+import kafka.server.checkpoints.LeaderEpochCheckpoint
+import kafka.server.epoch.Constants.{UNSUPPORTED_EPOCH, UNSUPPORTED_EPOCH_OFFSET}
+import kafka.utils.CoreUtils._
+import kafka.utils.{Logging}
+import org.apache.kafka.common.requests.EpochEndOffset
+import scala.collection.mutable.ListBuffer
+
+trait LeaderEpochCache {
+  def assignToLeo(leaderEpoch: Int)
+  def assign(leaderEpoch: Int, offset: Long)
+  def latestEpoch(): Int
+  def endOffsetFor(epoch: Int): Long
+  def clearLatest(offset: Long, retainMatchingOffset: Boolean = true)
+  def clearOldest(offset: Long, retainMatchingOffset: Boolean = true)
+  def clear()
+}
+
+object Constants {
+  val UNSUPPORTED_EPOCH_OFFSET = EpochEndOffset.UNDEFINED_OFFSET
+  val UNSUPPORTED_EPOCH = -1
+}
+
+/**
+  * Represents a cache of (LeaderEpoch => Offset) mappings for a particular replica.
+  *
+  * Leader Epoch = epoch assigned to each leader by the controller.
+  * Offset = offset of the first message in each epoch.
+  *
+  * @param leo a function that determines the log end offset
+  * @param checkpoint the checkpoint file
+  */
+class LeaderEpochFileCache(leo: () => LogOffsetMetadata, checkpoint: LeaderEpochCheckpoint) extends LeaderEpochCache with Logging {
+  private val lock = new ReentrantReadWriteLock()
+  private[epoch] var epochs = lock synchronized { ListBuffer(checkpoint.read(): _*) }
+
+  /**
+    * Assigns the supplied Leader Epoch to the current LEO
+    * Once the epoch is assigned it cannot be reassigned
+    *
+    * @param epoch
+    */
+  override def assignToLeo(epoch: Int) = {
+    assign(epoch, leo().messageOffset)
+  }
+
+  /**
+    * Assigns the supplied Leader Epoch to the supplied Offset
+    * Once the epoch is assigned it cannot be reassigned
+    *
+    * @param epoch
+    * @param offset
+    */
+  override def assign(epoch: Int, offset: Long): Unit = {
+    inWriteLock(lock) {
+      if (epoch >= 0 && epoch > latestEpoch()) {
+        epochs += EpochEntry(epoch, offset)
+        flush()
+      }
+    }
+  }
+
+  /**
+    * Returns the current Leader Epoch
+    *
+    * @return
+    */
+  override def latestEpoch(): Int = {
+    inReadLock(lock) {
+      if (epochs.isEmpty) UNSUPPORTED_EPOCH else epochs.last.epoch
+    }
+  }
+
+  /**
+    * Returns the End Offset for a requested Leader Epoch.
+    *
+    * This is defined as the start offset of the first Leader Epoch larger than the
+    * Leader Epoch requested, or else the Log End Offset if the latest epoch was requested.
+    *
+    * @param requestedEpoch
+    * @return offset
+    */
+  override def endOffsetFor(requestedEpoch: Int): Long = {
+    inReadLock(lock) {
+      val offset =
+      //Use LEO if current epoch, else the start offset of the first subsequent epoch
+        if (requestedEpoch == latestEpoch) {
+          leo().messageOffset
+        }
+        else {
+          val subsequentEpochs = epochs.filter(e => e.epoch > requestedEpoch)
```

Comment:
It seems that the method's Java doc is inconsistent with its behavior if requestedEpoch is < the first epoch in epochs. I am wondering if we should update the Java doc or comment in this code to explain this. I only realized it is intentional after reading this @junrao's comment in this pull request.